src/Pos/Delegation/Logic/Mempool.hs

```diff
@@ -177,8 +179,13 @@ processProxySKHeavy psk = do
         hasPskInDB <- isJust <$> GS.getPskByIssuer (Left $ pskIssuerPk psk)
         let rerevoke = isRevokePsk psk && not hasPskInDB
         producesCycle <- use dwProxySKPool >>= \pool ->
-            let eActions = map GS.pskToDlgEdgeAction pool
-            in lift $ detectCycleOnAddition (GS.withEActionsResolve eActions) psk
+            -- This is inefficient. Consider supporting this map
+            -- in-memory or changing mempool key to stakeholderId.
```

Comment:
Yes, it's certainly sad. Will you create an issue to improve it? In other components we store modifier together with mempool.